### ISSUE TITLE
Missing visibility annotations for QPs in initialize-globals.js

### ISSUE THREAD

**Issue by chrisklus** [phetsims/chipper#829]
From https://github.com/phetsims/joist/issues/593, we now have a flag `public` that can be added to query parameter schemas, which indicates that invalid values should fail gracefully and revert to default values.

@pixelzoom suggested I add this flag to the public query parameters in initialize-globals.js. Of the ~64 available query parameters, only 1 (`allowLinks`) is marked as public, and only 6 are marked as internal. The current annotations are also not standardized, so they are hard to find quickly.

I'd think that at least `screens` should also be marked as public since we advertise it via YouTube. I'm guessing several others should be, too. Not sure who to assign this to, so starting off with @ariel-phet.

**Comment by ariel-phet** [comment]
@samreid @zepumph can you look at these - I think it should be fairly quick. Basically I would say the following:

1. You should know pretty well which ones of these query parameters are needed to be public for iO and for a11y.

2. There are others (non iO/a11y related) that should be fairly obvious that we do not make public facing (I think something like `?phet-app` being a good example).

3. For any others that you are unsure of (for instance I believe we have one related to projector mode and such) - feel free to assign me to look over others.

**Comment by samreid** [comment]
I took a look and it seems like everything marked with `@memberOf PhetQueryParameters` should be labeled as `public: true` (in phet brand and phet-io brand).  This would be good to double check with @zepumph, so I'll leave it assigned to him for consideration.  Other notes and questions that came up during my review:

* `audioVolume` is listed as deprecated, but I don't see a replacement.  @jbphet is there a proposed replacement?  If so, should it be public-facing?  Co-assigning @jbphet for this part.
* Should `colorProfile` be public?  If not, then what is it for?
* I'm not familiar with several Query Parameters such as `qrCode`. Maybe @zepumph knows or we can ask other developers to chime in.
* We already have another key `public`, which means that our schema permits the following:

```js
public:true 
private: true
```

Should we change this to an enumeration like `visibility: ('public','private','internal')`?  Or should we add assertions that something cannot be both public and private?  Co-assigning @chrisklus and @pixelzoom to comment on this part.

**Comment by chrisklus** [comment]
+1 for changing the key to something like `visibility`. Do you know if the difference between private and internal in our documentation is just that internal is only meant for phet team members but is not as strict as requiring `phetTeamMember` in localStorage like private does? And if so, do we like that functionality of internal, or would just public and private be sufficient?

This key could replace annotations like `"Intended for internal use only"`, which would be easier to parse when skimming QP files and be functional instead of only serving as documentation.

**Comment by pixelzoom** [comment]
I had no idea that there was a 'private' property.  Neither 'public' nor 'private' are described in the documentation for `const TYPES`, and they should be.   And shouldn't 'private' be included in `const validateSchema`?

That said, I don't understand the need for `visibility: ('public'|'private'|'internal')`.  What is the difference between 'private' and 'internal'?  Aren't they the same thing?

**Comment by pixelzoom** [comment]
Oh, I see... 

'public' = public facing
'internal' = intended for use only by PhET team members 
'private' = protected from use by non-PhET team members (query parameter is ignore)

This is confusing.  I think it would be better to just have 'public' and 'private', and have a separate property (applicable to 'private' query parameters) that indicates whether to ignore the query parameter if it's provide.

**Comment by pixelzoom** [comment]
I'd also be OK with `visibility: 'public'|'private'`.   But I'm not crazy about `visibility: 'public'|'private'|'internal'`.  But I might be convinced if the values made the semantics clear.

**Comment by jbphet** [comment]
> * `audioVolume` is listed as deprecated, but I don't see a replacement.  @jbphet is there a proposed replacement?  If so, should it be public-facing?

The `audioVolume` parameter was originally implemented to support having sound disabled at sim launch time so that anyone doing fuzz testing wasn't hit with a barrage of sounds.  This is supported in sims that use tambo as the sound library using the `sound=disabled` query parameter.  And, though this support does exist for disabling sound during fuzz testing, it's not as much of an issue now that Chrome and other browsers have adopted a policy where sound can't be played until the user interacts with the page in some way.  For this particular use case (i.e. fuzz testing), it's not necessary to make the `sound` parameter public facing.  However, I suspect that at some point we **will** make it public so that teachers can disable sound using a query parameter if they need to keep the classroom quiet.  So far, we haven't received any such requests that I'm aware of, so I'm not sure of the urgency of making this public.  That seems like a question for @ariel-phet and/or @kathy-phet to me.

**Comment by chrisklus** [comment]
@pixelzoom said:

>This is confusing. I think it would be better to just have 'public' and 'private', and have a separate property (applicable to 'private' query parameters) that indicates whether to ignore the query parameter if it's provided.

Agreed that it's confusing, and that we should use just 'public' and 'private'. But why wouldn't we leave the functionality of 'public' and 'private' as is, so they both serve a purpose? If we move the "ignore the query parameter if it's provided" functionality to a different key, then 'private' is confusing to me because it's only documentation. Are there QPs that we want marked as 'private' but still allowed to be used by non PhET team members?

**Comment by pixelzoom** [comment]
> why wouldn't we leave the functionality of 'public' and 'private' as is, so they both serve a purpose?

It sounds like you're suggesting that 'internal' and 'private' should be combined.  If that were the case, then all PhET team members would be forced to use localStorage in order to use query parameters that aren't public-facing.   That will not fly.  As I recall, non-developers were not excited about having to deal with localStorage, and went along with the solution only because they expect not to have to use it.

**Comment by samreid** [comment]
Self unassigning until @zepumph has a chance to take a look.  @zepumph please re-assign me after you comment.

**Comment by zepumph** [comment]
For IO I agree with @samreid, all marked with `@memberof` annotations are rendered into phet-io jsdoc and are public parameters as it pertains to phet-io.

For a11y, `enhancedSoundIinitiallyEnabled` seems like there is potential for public, but I am no authority, perhaps @jbphet can weigh in on that one too. 

`visibility` is verbose but clear, as long as all available allowed keys are well documented. I also agree with @pixelzoom that we should not consolidate `internal` and `private`, and perhaps should rename `private` to be more explicit, like `secret` or `team-only` or something.

Back to you @samreid.

**Comment by jbphet** [comment]
> For a11y, enhancedSoundIinitiallyEnabled seems like there is potential for public, but I am no authority, perhaps @jbphet can weigh in on that one too.

Yes, that once should be public.  I added a line to the doc that I saw elsewhere in the file that says, "This query parameter is public facing."  Does anything else need to be done for this?

**Comment by samreid** [comment]
Summarizing the current status:

initialize-globals public:
* allowLinks
* homeScreen
* initialScreen
* screens
* supportsPanAndZoom

initialize-globals private:
* showAnswers

**Comment by samreid** [comment]
> For IO I agree with @samreid, all marked with @memberof annotations are rendered into phet-io jsdoc and are public parameters as it pertains to phet-io.

In the commit, I also marked memberof things as public:
* sound
* phetioEmitHighFrequencyEvents
* phetioEmitStates
* phetioEmitStatesInterval
* phetioConsoleLog
* phetioStandalone

**Comment by samreid** [comment]
> For a11y, enhancedSoundIinitiallyEnabled seems like there is potential for public

I marked that `public: true` in the commit.  

There are proposals in this issue for combining public/private into visibility, but there is disagreement and it is not clear to be worth the effort.  I'm satisfied with the current status of this issue, but think I should leave it up to commenters whether they would like other steps to be taken before closing (perhaps in a side issue).  @pixelzoom @chrisklus or @zepumph what do you recommend?

**Comment by pixelzoom** [comment]
I'm OK with `public`.   That said, I would be very happy if `public` (and `private`, and `protected`) became JavaScript keywords in the future, if JavaScript were to provide language support for visibility.

**Comment by zepumph** [comment]
I feel alright with the current implementation. Anyone feel free to re-open.
